### ISSUE TITLE
Startup database load not working properly with NumberFormatException

### ISSUE THREAD

**Issue by Boca13** [LINBIT/linstor-server#248]
Without any intentional changes on my side, my Linstor controller has stopped loading properly. The issue is that after startup, everything seems empty: no volumes or resources, no satellites, no storage pools... I've checked the etcd data, but the information seems to be there.

An error occurs at startup, apparently related to Java not being able to parse a value from the database as Long. I've checked those lines of code, but I'm not sure what's happening. It seems to me that it happens when loading flags from the "LAYER_DRBD_RESOURCES" table/key. I can't find any values for such flags in my etcd dump, but my understanding is limited.

https://github.com/LINBIT/linstor-server/blob/6811995e1cf099150eb8e112390857e7b846db46/controller/src/main/java/com/linbit/linstor/core/objects/DrbdLayerETCDDriver.java#L529

I'm using Linstor server 1.12.3 with Piraeus (1.5).

Excerpt from linstor server log:

```
12:18:28.763 [Main] INFO  LINSTOR/Controller - SYSTEM - SpaceTracking using ETCD driver
12:18:28.765 [Main] INFO  LINSTOR/Controller - SYSTEM - SpaceTrackingService: Instance added as a system service
12:18:28.767 [Main] INFO  LINSTOR/Controller - SYSTEM - Starting service instance 'TimerEventService' of type TimerEventService
12:18:28.768 [Main] INFO  LINSTOR/Controller - SYSTEM - Initializing the etcd database
12:18:30.751 [Main] INFO  LINSTOR/Controller - SYSTEM - Starting service instance 'ETCDDatabaseService' of type ETCDDatabaseService
12:18:30.753 [Main] INFO  LINSTOR/Controller - SYSTEM - Loading security objects
12:18:31.265 [Main] INFO  LINSTOR/Controller - SYSTEM - Current security level is NO_SECURITY
12:18:31.631 [Main] INFO  LINSTOR/Controller - SYSTEM - Core objects load from database is in progress
12:18:38.348 [Main] ERROR LINSTOR/Controller - SYSTEM - null [Report number 61029C8C-00000-000000]
12:18:38.355 [Main] INFO  LINSTOR/Controller - SYSTEM - Starting service instance 'TaskScheduleService' of type TaskScheduleService
12:18:38.356 [Main] INFO  LINSTOR/Controller - SYSTEM - Initializing network communications services
12:18:38.358 [TaskScheduleService] INFO  LINSTOR/Controller - SYSTEM - LogArchive: Running log archive on directory: /var/log/linstor-controller
12:18:38.366 [TaskScheduleService] INFO  LINSTOR/Controller - SYSTEM - LogArchive: No logs to archive.
12:18:39.337 [Main] WARN  LINSTOR/Controller - SYSTEM - NetComService: Connector DebugSslConnector: Binding the socket to the IPv6 anylocal address failed, attempting fallback to IPv4
12:18:39.341 [Main] INFO  LINSTOR/Controller - SYSTEM - Created network communication service 'DebugSslConnector', bound to ::0:3373
12:18:39.343 [Main] INFO  LINSTOR/Controller - SYSTEM - Created network communication service 'PlainConnector'
12:18:39.568 [Main] INFO  LINSTOR/Controller - SYSTEM - Created network communication service 'SslConnector'
12:18:39.631 [Main] ERROR LINSTOR/Controller - SYSTEM - The current scope has already been entered [Report number 61029C8C-00000-000001]

12:18:39.632 [Main] INFO  LINSTOR/Controller - SYSTEM - No known nodes.
```
```
12:20:39.671 [SpaceTrackingService] INFO  LINSTOR/Controller - SYSTEM - SpaceTracking: Found no satellite nodes
12:20:39.673 [SpaceTrackingService] INFO  LINSTOR/Controller - SYSTEM - SpaceTracking: Aggregate capacity is 0 kiB
```
Full error report:
```
ERROR REPORT 61029C8C-00000-000000

============================================================

Application:                        LINBIT�� LINSTOR
Module:                             Controller
Version:                            1.12.3
Build ID:                           d4e2cbfcb3819600208b3e4849e9efa6ddb50a52
Build time:                         2021-05-07T06:20:48+00:00
Error time:                         2021-07-29 12:18:38
Node:                               piraeus-op-cs-controller-67776754d7-wk7d5

============================================================

Reported error:
===============

Category:                           RuntimeException
Class name:                         NumberFormatException
Class canonical name:               java.lang.NumberFormatException
Generated at:                       Method 'parseLong', Source file 'Long.java', Line #655

Error message:                      null

Call backtrace:

    Method                                   Native Class:Line number
    parseLong                                N      java.lang.Long:655
    parseLong                                N      java.lang.Long:817
    load                                     N      com.linbit.linstor.core.objects.DrbdLayerETCDDriver:529
    loadLayerData                            N      com.linbit.linstor.dbdrivers.DatabaseLoader:713
    loadLayerObects                          N      com.linbit.linstor.dbdrivers.DatabaseLoader:617
    loadAll                                  N      com.linbit.linstor.dbdrivers.DatabaseLoader:546
    loadCoreObjects                          N      com.linbit.linstor.core.DbDataInitializer:176
    initialize                               N      com.linbit.linstor.core.DbDataInitializer:108
    startSystemServices                      N      com.linbit.linstor.core.ApplicationLifecycleManager:87
    start                                    N      com.linbit.linstor.core.Controller:343
    main                                     N      com.linbit.linstor.core.Controller:564


END OF ERROR REPORT.
```
Any ideas? Maybe something is missing from my database?

Thank you.

**Comment by Boca13** [comment]
I tried with Linstor server 1.13.0 and I got the same error, just with some line number shifts:
```
ERROR REPORT 6103C019-00000-000000

============================================================

Application:                        LINBIT�� LINSTOR
Module:                             Controller
Version:                            1.13.0
Build ID:                           37c02e20aa52f26ef28ce4464925d9e53327171c
Build time:                         2021-06-21T06:45:49+00:00
Error time:                         2021-07-30 09:02:32
Node:                               piraeus-op-cs-controller-7dfdb75889-qqdrz

============================================================

Reported error:
===============

Category:                           RuntimeException
Class name:                         NumberFormatException
Class canonical name:               java.lang.NumberFormatException
Generated at:                       Method 'parseLong', Source file 'Long.java', Line #655

Error message:                      null

Call backtrace:

    Method                                   Native Class:Line number
    parseLong                                N      java.lang.Long:655
    parseLong                                N      java.lang.Long:817
    load                                     N      com.linbit.linstor.core.objects.DrbdLayerETCDDriver:529
    loadLayerData                            N      com.linbit.linstor.dbdrivers.DatabaseLoader:732
    loadLayerObects                          N      com.linbit.linstor.dbdrivers.DatabaseLoader:636
    loadAll                                  N      com.linbit.linstor.dbdrivers.DatabaseLoader:561
    loadCoreObjects                          N      com.linbit.linstor.core.DbDataInitializer:176
    initialize                               N      com.linbit.linstor.core.DbDataInitializer:108
    startSystemServices                      N      com.linbit.linstor.core.ApplicationLifecycleManager:87
    start                                    N      com.linbit.linstor.core.Controller:343
    main                                     N      com.linbit.linstor.core.Controller:564


END OF ERROR REPORT.
```

**Comment by rp-** [comment]
could you provide a dump of you etcd database?

**Comment by Boca13** [comment]
Sure:
- [text dump](https://assets.libnamic.com/linstor.txt)
- [db **snapshot](https://assets.libnamic.com/linstor.db)**

Thank you so much.

**Comment by ghernadi** [comment]
> An error occurs at startup, apparently related to Java not being able to parse a value from the database as Long. I've checked those lines of code, but I'm not sure what's happening. It seems to me that it happens when loading flags from the "LAYER_DRBD_RESOURCES" table/key. I can't find any values for such flags in my etcd dump, but my understanding is limited.

Your understanding is quite correct. If you take a look at the "text dump" you provided, the first few lines belong to the same DRBD layer data with the id `0`:
```
/LINSTOR/LAYER_DRBD_RESOURCES/0/AL_STRIPES
1
/LINSTOR/LAYER_DRBD_RESOURCES/0/AL_STRIPE_SIZE
32
/LINSTOR/LAYER_DRBD_RESOURCES/0/NODE_ID
0
/LINSTOR/LAYER_DRBD_RESOURCES/0/PEER_SLOTS
7
```
However, there is actually an entry missing, namely `/LINSTOR/LAYER_DRBD_RESOURCES/0/FLAGS` with some value (might be 0, but I can NOT recommend to assume 0 as default)

> Without any intentional changes on my side [...]

Might not be intentional or not from your side, but something had to change the ETCD datastore, as I highly doubt that ETCD would precisely remove all `/LINSTOR/LAYER*/*/FLAGS` entries from your datastore.

> [...] my Linstor controller has stopped loading properly. The issue is that after startup, everything seems empty: no volumes or resources, no satellites, no storage pools...

Yes, that is more or less a known bug. I agree that the controller should not even start in case of a DatabaseException (whereas exceptions from ETCD are in the same category). Having an "up and running" controller which appears to have no data is quite confusing, I agree.

>  I've checked the etcd data, but the information seems to be there.

Most of the data seems to be there, but as mentioned above some data is simply missing.

Sorry, but right now I do not really see what we can do here. If you have a backup you might want to use that to restore those `/LINSTOR/..../FLAGS` keys + values, but there is also no guarantee that this will "just work". Fully restoring your backup might be your best bet.

**Comment by Boca13** [comment]
Thank you so much for your detailed answer, @ghernadi . I took the `member/snap/db` from an old backup and restored it into an auxiliary etcd pod. Then I checked the FLAGS keys and realized that all flags were 0 except of one resource, for which they were 2.

I created a script that would just set the keys to their right values according to that backup and it worked! I felt like it was a bit safer than just restoring the old backup, so I wouldn't lose any new data. Luckily, there weren't any new resources with unknown flags that we had had to guess.

It was indeed a weird incident, but now I know that if something like this happened again, I could just restore a recent backup right away. Hopefully, if the controller changes its behavior in next versions to not start at all when a DatabaseException occurs, we will realize similar problems earlier.

Thank you again for your support and maintaining this great project!
